Global pointer arithmetics and pointer distance in global memory

fuchsto opened the pull request:
Introduces correct random access on global address space. `GlobPtr` now supports iteration across unit memory spaces for balanced, unbalanced and dynamic memory spaces.

See:
https://github.com/dash-project/dash/blob/a26b27ea142449d0a9f4beb4d201f7a912aad002/dash/test/memory/GlobStaticMemTest.cc#L33

The full story is discussed in https://github.com/dash-project/dash/pull/333

Before the changes in this PR, `GlobPtr` continued iteration past the local memory space instead of the subsequent unit's address range.
Some places in DASH rely on the previous "relaxed" semantics of global pointer arithmetics, so some additional adjustments will soon be pushed.

Update: Done and ready for review now

fuchsto committed:
Fixing random-access on global address range in GlobPtr

fuchsto committed:
Merge branch 'development' of github.com:dash-project/dash into feat-globptr

fuchsto committed:
Integrating strict random-access semantics on global address range

fuchsto committed:
Fixed nullptr check in GlobStaticMem

rkowalewski commented on dash/include/dash/GlobPtr.h:
```diff
@@ -543,9 +488,57 @@ class GlobPtr
+  }
+
+private:
+  void increment(index_type offs) {
```
This sequence looks really ugly. I would suggest to modify the code  for better readabality with appropriate variables or even a small macros. Something like this:
~~~cpp
//Local Variables
auto const & raddr = _rbegin_gptr.addr_or_offs.offset;
auto const lpos = _raddr / sizeof(value_type);
// Alternatively
#define _LPOS (addr) (addr / sizeof(element_type))
auto const lpos = _LPOS(r_addr);
~~~

I remember we decided to omit `dart_gptr_incaddr` since it does currently nothing but incrementing address and offset. This might be actually problematic IMHO. Will write something in #333

rkowalewski commented on dash/include/dash/memory/GlobHeapMem.h:
```diff
@@ -429,6 +429,23 @@ class GlobHeapMem
     return unit_local_size;
   }
 
```
See my comments to `GlobPtr`. The same applies here. Maybe we can centralize the pointer arithmetic stuff.

fuchsto replied:
What's wrong with `GlobHeapMem::local_size`? It's fundamentally different from the implementation in any other memory space type, so why should it be centralized?

rkowalewski replied:
Öhm actually the comment refers to the wrong line. What I mean is `GlobHeapMem::distance`. And the problem is not the replicated logic itself. It is clear that `GlobHeapMem::distance` is different from that in `GlobStaticMem`. What I mean is the `gptr.addr_or_offset.offset / sizeof(element_type)` which is repeated in at least 3 or 4 files. We could provide a simple macro for this.

fuchsto replied:
Ah, yes. Actually I want to get rid of the division ops somehow and multiply the byte-based offsets instead. As we know from ParComp, division is horribly expensive and pointer increments should have as little overhead as possible.

devreal replied:
It would be interesting to see whether this ends up to be a `div` in the object code. I'd expect the compiler to optimize it since `sizeof(element_type)` is a compile-time constant.

fuchsto replied:
... and `sizeof(element_type)` is a power of two, so it should become a left-shift, right.

fuchsto replied:
... right-shift I mean

rkowalewski commented on dash/include/dash/memory/GlobStaticMem.h:
```diff
@@ -294,6 +288,23 @@ class GlobStaticMem
     return _nlelem;
   }
 
+  index_type distance(const pointer & gbegin, const pointer & gend) const {
```
Even more duplicated stuff here.

fuchsto replied:
The calculation of global pointer distance depends on `Glob*Mem` as discussed in #333 so it cannot be implemented as a single independent function.
For comparison, see `dash::GlobUnitMem`.

We could specialize `dash::distance` for `GlobPtr` and its dependent `Glob*Mem` and define `GlobPtr::operator-(const self_t &)` as `dash::distance(*this, other)`.

devreal commented on dash/include/dash/GlobPtr.h:
```diff
@@ -104,12 +110,24 @@ class GlobPtr
+    dart_gptr_t         gptr,
+    index_type          rindex = 0)
+  : _rbegin_gptr(gptr)
   , _mem_space(reinterpret_cast<const MemorySpace *>(mem_space))
```
What is this `reinterpret_cast` good for?

fuchsto replied:
Ah, the argument type should be `const MemSpaceT`. Assignment and copy from other GlobPtr with different memory space type is essential to convert `GlobPtr<T>` to `GlobPtr <const T>`

devreal commented on dash/include/dash/GlobPtr.h:
```diff
@@ -290,47 +305,41 @@ class GlobPtr
+             ? ( _rbegin_gptr.addr_or_offs.offset
+                 - rhs._rbegin_gptr.addr_or_offs.offset )
+               / sizeof(value_type)
+             : _mem_space->distance(rhs, *this) );
```
correct ordering or arguments? Better: `_mem_space->distance(*this, rhs)`

fuchsto replied:
No, because `*this - rhs = distance(rhs, *this)`

devreal replied:
You're right, my mistake.

devreal reviewed (commented):
Comments inline. Looks good in general :+1:

fuchsto commented:
@rkowalewski I eliminated duplicate code using specialization of `dash::distance` for global pointer types and their dependent memory space type.
The increment/decrement operations in `GlobPtr` have been cleaned up a bit.

fuchsto commented:
@rkowalewski @devreal 

Changes following your requests:

- I managed to eliminate `GlobHeapMem.distance` and `GlobStaticMem.distance` (and thus their duplicate implementations) using specialization of `dash::distance` for global pointer types and their dependent memory space type.
- the increment/decrement operations in `GlobPtr` have been cleaned up a bit.

Also:

- updated other specializations of `dash::distance` accordingly
- updated doxygen documentation and removed resolved TODOs

fuchsto committed:
Added dash::distance(GlobPtr,GlobPtr), updated documentation

fuchsto commented:
@rkowalewski @devreal 

Changes following your requests:

- I managed to eliminate `GlobHeapMem.distance` and `GlobStaticMem.distance` (and thus their duplicate implementations) using specialization of `dash::distance` for global pointer types and their dependent memory space type. (see https://github.com/dash-project/dash/blob/22b000777dd5fb444e0ccb9913151876fce8189b/dash/include/dash/GlobPtr.h#L793)
- the increment/decrement operations in `GlobPtr` have been cleaned up a bit.

Also:

- updated other specializations of `dash::distance` accordingly
- updated doxygen documentation and removed resolved TODOs

fuchsto committed:
Merge branch 'development' into feat-globptr

fuchsto committed:
Updated dash::distance(dart_gptr_t,dart_gptr_t), documentation

fuchsto committed:
Type constraints in dash::distance(GlobPtr,GlobPtr)

fuchsto committed:
Type constraints in dash::distance(GlobPtr,GlobPtr)

fuchsto committed:
Extended tests of global pointer random access

fuchsto committed:
Updated GlobPtr.local()

fuchsto commented:
@rkowalewski @devreal 

Changes following your requests:

- I managed to eliminate `GlobHeapMem.distance` and `GlobStaticMem.distance` (and thus their duplicate implementations) using specialization of `dash::distance` for global pointer types and their dependent memory space type. (see https://github.com/dash-project/dash/blob/22b000777dd5fb444e0ccb9913151876fce8189b/dash/include/dash/GlobPtr.h#L793)
- the increment/decrement operations in `GlobPtr` have been cleaned up a bit.

Also:

- updated other specializations of `dash::distance` accordingly
- updated doxygen documentation and removed resolved TODOs
- refactored `GlobBucketIter` to specialization of `GlobPtr` so the global pointer concept now is maintained for dynamic memory spaces

fuchsto committed:
Refactored GlobBucketIter to specialization of GlobPtr

fuchsto committed:
Merge branch 'feat-globptr' of github.com:dash-project/dash into feat-globptr

fuchsto committed:
Fixed compiler errors

fuchsto committed:
Minor, cleanup of internal typedefs

devreal reviewed (approved):
Looks good to me :)
